src/comparator.rs

```diff
@@ -15,164 +15,103 @@
 use std::cmp::Ordering;
-use std::convert::TryInto;
 use std::ffi::CString;
-use std::mem::size_of;
 use std::slice;
 
 pub type CompareFn = dyn Fn(&[u8], &[u8]) -> Ordering;
 
-// Use u64 as the timestamp. This is based on two reasons:
-// 1. Follows the logic of [BytewiseComparatorWithU64Ts](https://github.com/facebook/rocksdb/blob/3db030d7ee1b887ce818ec6f6a8d10949f9e9a22/util/comparator.cc#L238)
-// 2. u64 is the return type of [Duration::as_secs()](https://doc.rust-lang.org/nightly/std/time/struct.Duration.html#method.as_secs)
-fn strip_timestamp_from_user_key(user_key: &[u8], ts_sz: usize) -> &[u8] {
-    &user_key[..user_key.len() - ts_sz]
-}
+pub type CompareTsFn = dyn Fn(&[u8], &[u8]) -> Ordering;
 
-fn extract_timestamp_from_user_key(user_key: &[u8], ts_sz: usize) -> &[u8] {
-    &user_key[user_key.len() - ts_sz..]
-}
+pub type CompareWithoutTsFn = dyn Fn(&[u8], bool, &[u8], bool) -> Ordering;
 
-// Caller should ensure the pointer is valid and has at least 8 bytes,
-// As the slice::from_raw_parts does in compare_ts_callback
-#[inline]
-fn decode_timestamp(ptr: &[u8]) -> u64 {
-    u64::from_be_bytes(ptr[..8].try_into().unwrap())
+pub struct ComparatorCallback {
+    pub name: CString,
+    pub compare_fn: Box<CompareFn>,
 }
 
-fn compare_ts(a: &[u8], b: &[u8]) -> c_int {
-    let a = decode_timestamp(a);
-    let b = decode_timestamp(b);
-    match a.cmp(&b) {
-        Ordering::Less => -1,
-        Ordering::Equal => 0,
-        Ordering::Greater => 1,
+impl ComparatorCallback {
+    pub unsafe extern "C" fn destructor_callback(raw_cb: *mut c_void) {
+        drop(Box::from_raw(raw_cb as *mut Self));
     }
-}
 
-pub struct ComparatorCallback {
-    pub name: CString,
-    pub f: Box<CompareFn>,
-}
+    pub unsafe extern "C" fn name_callback(raw_cb: *mut c_void) -> *const c_char {
+        let cb: &mut Self = &mut *(raw_cb as *mut Self);
+        let ptr = cb.name.as_ptr();
+        ptr as *const c_char
+    }
 
-pub unsafe extern "C" fn destructor_callback(raw_cb: *mut c_void) {
-    drop(Box::from_raw(raw_cb as *mut ComparatorCallback));
+    pub unsafe extern "C" fn compare_callback(
+        raw_cb: *mut c_void,
+        a_raw: *const c_char,
+        a_len: size_t,
+        b_raw: *const c_char,
+        b_len: size_t,
+    ) -> c_int {
+        let cb: &mut Self = &mut *(raw_cb as *mut Self);
+        let a: &[u8] = slice::from_raw_parts(a_raw as *const u8, a_len);
+        let b: &[u8] = slice::from_raw_parts(b_raw as *const u8, b_len);
+        (cb.compare_fn)(a, b) as c_int
+    }
 }
 
-pub unsafe extern "C" fn name_callback(raw_cb: *mut c_void) -> *const c_char {
-    let cb: &mut ComparatorCallback = &mut *(raw_cb as *mut ComparatorCallback);
-    let ptr = cb.name.as_ptr();
-    ptr as *const c_char
+pub struct ComparatorWithTsCallback {
+    pub name: CString,
+    pub compare_fn: Box<CompareFn>,
+    pub compare_ts_fn: Box<CompareTsFn>,
+    pub compare_without_ts_fn: Box<CompareWithoutTsFn>,
 }
```

Comment:
Now we have two structs here:

- `ComparatorCallback`, which is used in `Options::set_comparator`
- `ComparatorWithTsCallback`, which is used in `Options::set_comparator_with_ts`